### PR TITLE
Java: Add `BZMPOP` command

### PR DESCRIPTION
**Pull request by Yury-Fridlyand**
https://redis.io/commands/bzmpop/

ref: #188 and #181

### CODE REVIEW

File: java/client/src/main/java/glide/api/models/BaseTransaction.java
Comment (jonathanl-bq) [inline java/client/src/main/java/glide/api/models/BaseTransaction.java:1778]: This sentence needs to be reworded because it doesn't make sense as is.

File: java/client/src/main/java/glide/api/models/BaseTransaction.java
Comment (jonathanl-bq) [inline java/client/src/main/java/glide/api/models/BaseTransaction.java:1779]: Why is BZMPOP in all caps when zmpop isn't?

File: java/client/src/main/java/glide/api/models/BaseTransaction.java
Comment (jonathanl-bq) [inline java/client/src/main/java/glide/api/models/BaseTransaction.java:1795]: This sentence also needs to be reworded because it doesn't make sense.

File: java/client/src/main/java/glide/api/models/BaseTransaction.java
Comment (jonathanl-bq) [inline java/client/src/main/java/glide/api/models/BaseTransaction.java:1764]: Same feedback about documentation here.

File: java/client/src/main/java/glide/api/commands/SortedSetBaseCommands.java
Comment (jonathanl-bq) [inline java/client/src/main/java/glide/api/commands/SortedSetBaseCommands.java:491]: Same documentation feedback here.

File: java/client/src/main/java/glide/api/commands/SortedSetBaseCommands.java
Comment (jonathanl-bq) [inline java/client/src/main/java/glide/api/commands/SortedSetBaseCommands.java:524]: Same documentation feedback here.

**Comment by Yury-Fridlyand** [comment]
Updated docs, please, re-read.

File: java/client/src/main/java/glide/api/models/BaseTransaction.java
Comment (jonathanl-bq) [inline java/client/src/main/java/glide/api/models/BaseTransaction.java:1793]: This sentence still doesn't make sense grammatically. Do you mean: "The data could be interpreted as [key, {element: score}]"? It might also be easier to just delete this line entirely, since the above line already explains what is returned.

File: java/client/src/main/java/glide/api/models/BaseTransaction.java
Comment (jonathanl-bq) [inline java/client/src/main/java/glide/api/models/BaseTransaction.java:1779]:
```suggestion
     * Members with the highest/lowest score are popped from the first sorted set that is non-empty, with
```

File: java/client/src/main/java/glide/api/models/BaseTransaction.java
Comment (jonathanl-bq) [inline java/client/src/main/java/glide/api/models/BaseTransaction.java:1759]: Should this also be "were popped" like in the other bzmpop function below?
```suggestion
     *     from which an elements were popped, and a member-score <code>Map</code>.<br>
```

File: java/client/src/main/java/glide/api/models/BaseTransaction.java
Comment (jonathanl-bq) [inline java/client/src/main/java/glide/api/models/BaseTransaction.java:1777]: The intention of this is a bit clearer, but it still doesn't make sense. Is this what you meant?
```suggestion
     * Blocks the connection until it pops one or more member-score pairs from the 
     * first non-empty sorted set in the provided list of key names.<br>
```

File: java/client/src/main/java/glide/api/models/BaseTransaction.java
Comment (Yury-Fridlyand) [reply java/client/src/main/java/glide/api/models/BaseTransaction.java:1759]: No - it returns only one element

File: java/client/src/main/java/glide/api/models/BaseTransaction.java
Comment (Yury-Fridlyand) [reply java/client/src/main/java/glide/api/models/BaseTransaction.java]: I initially copied from another doc/PR. Removed.

File: java/client/src/main/java/glide/api/commands/SortedSetBaseCommands.java
Comment (SanHalacogluImproving) [inline java/client/src/main/java/glide/api/commands/SortedSetBaseCommands.java:464]: Blocks the connection until it pops and returns a member-score pair from the sorted sets stored at the specified <code>keys</code>. The sorted sets are checked in the order. 

This should be enough.

File: java/client/src/main/java/glide/api/commands/SortedSetBaseCommands.java
Comment (SanHalacogluImproving) [inline java/client/src/main/java/glide/api/commands/SortedSetBaseCommands.java:466]: Remove this line.

File: java/client/src/main/java/glide/api/commands/SortedSetBaseCommands.java
Comment (SanHalacogluImproving) [inline java/client/src/main/java/glide/api/commands/SortedSetBaseCommands.java:468]: You could add this to the description at the top.

File: java/client/src/main/java/glide/api/commands/SortedSetBaseCommands.java
Comment (SanHalacogluImproving) [inline java/client/src/main/java/glide/api/commands/SortedSetBaseCommands.java:471]: Did Andrew say to remove these?

File: java/client/src/main/java/glide/api/commands/SortedSetBaseCommands.java
Comment (SanHalacogluImproving) [inline java/client/src/main/java/glide/api/commands/SortedSetBaseCommands.java:477]: You should talk about the possible modifiers. Min and Max.

File: java/client/src/main/java/glide/api/commands/SortedSetBaseCommands.java
Comment (SanHalacogluImproving) [inline java/client/src/main/java/glide/api/commands/SortedSetBaseCommands.java:478]: Again we should decide on member vs element. Pls change all elements to members if you are sticking with that.

File: java/client/src/main/java/glide/api/commands/SortedSetBaseCommands.java
Comment (SanHalacogluImproving) [inline java/client/src/main/java/glide/api/commands/SortedSetBaseCommands.java:493]: Comment as above.

File: java/client/src/main/java/glide/api/commands/SortedSetBaseCommands.java
Comment (SanHalacogluImproving) [inline java/client/src/main/java/glide/api/commands/SortedSetBaseCommands.java:489]: Why is the return type a map but not for BZMPOPMAX? Seems confusing to have different return types for similar commands.

File: java/client/src/main/java/glide/api/models/BaseTransaction.java
Comment (SanHalacogluImproving) [inline java/client/src/main/java/glide/api/models/BaseTransaction.java:1743]: Comments as above.

File: java/client/src/main/java/glide/api/models/BaseTransaction.java
Comment (SanHalacogluImproving) [inline java/client/src/main/java/glide/api/models/BaseTransaction.java:1775]: Comments as above.

File: java/client/src/main/java/glide/api/models/commands/ZmpopOptions.java
Comment (SanHalacogluImproving) [inline java/client/src/main/java/glide/api/models/commands/ZmpopOptions.java:13]: Is this the only command that uses Count if not I propose you put it it in the SortedSetBaseCommands interface.

File: java/integTest/src/test/java/glide/SharedCommandTests.java
Comment (SanHalacogluImproving) [inline java/integTest/src/test/java/glide/SharedCommandTests.java:1296]: These should be tested with the correct order after Jonathan's LinkedHashMap PR gets merged, if not seems a bit pointless to test via max and min if set.length < count as we can't preserve order. So you probably should test where set.length > count to actually test the functionality.

File: java/client/src/main/java/glide/api/commands/SortedSetBaseCommands.java
Comment (Yury-Fridlyand) [reply java/client/src/main/java/glide/api/commands/SortedSetBaseCommands.java:471]: Will remove/update for all blocking commands (few of them already merged)
I copied this from python client - so we need to update it too in that case

File: java/client/src/main/java/glide/api/commands/SortedSetBaseCommands.java
Comment (Yury-Fridlyand) [reply java/client/src/main/java/glide/api/commands/SortedSetBaseCommands.java:478]: Those are array elements, not zset elements

File: java/client/src/main/java/glide/api/commands/SortedSetBaseCommands.java
Comment (Yury-Fridlyand) [reply java/client/src/main/java/glide/api/commands/SortedSetBaseCommands.java:489]: Ask redis, not me

File: java/integTest/src/test/java/glide/SharedCommandTests.java
Comment (Yury-Fridlyand) [reply java/integTest/src/test/java/glide/SharedCommandTests.java:1296]: Good point, will keep this unresolved until then

**Review by jonathanl-bq — APPROVED** [review]

**Comment by Yury-Fridlyand** [comment]
Unblocked

File: java/integTest/src/test/java/glide/SharedCommandTests.java
Comment (Yury-Fridlyand) [reply java/integTest/src/test/java/glide/SharedCommandTests.java]: Added

File: java/client/src/main/java/glide/api/models/commands/ZmpopOptions.java
Comment (acarbonetto) [inline java/client/src/main/java/glide/api/models/commands/ZmpopOptions.java:11]: why not just make this an enum like `ConditionalChange` instead of defining it only for ZMPOP?

File: java/client/src/main/java/glide/api/commands/SortedSetBaseCommands.java
Comment (acarbonetto) [inline java/client/src/main/java/glide/api/commands/SortedSetBaseCommands.java:910]: nit: remove whitespace

File: java/client/src/test/java/glide/api/RedisClientTest.java
Comment (acarbonetto) [inline java/client/src/test/java/glide/api/RedisClientTest.java:2115]: this should be a 2-element array

File: java/client/src/test/java/glide/api/RedisClientTest.java
Comment (acarbonetto) [inline java/client/src/test/java/glide/api/RedisClientTest.java:2142]: this should be a 2-element array

File: java/client/src/test/java/glide/api/models/TransactionTests.java
Comment (acarbonetto) [inline java/client/src/test/java/glide/api/models/TransactionTests.java:462]: merge conflict?

File: java/integTest/src/test/java/glide/SharedCommandTests.java
Comment (acarbonetto) [inline java/integTest/src/test/java/glide/SharedCommandTests.java:2406]: not sure this needs to be called twice

File: java/client/src/main/java/glide/api/BaseClient.java
Comment (acarbonetto) [inline java/client/src/main/java/glide/api/BaseClient.java:1081]: Can we make keys come before timeout?  
Generally, we put all the options together and then we could get away with a single function signature.

Comment (Yury-Fridlyand) [reply java/client/src/main/java/glide/api/BaseClient.java:1081]: I also dislike it, but in redis api, timeout is the first argument - we follow that
in `bzpopmax` it is the last...

File: java/client/src/main/java/glide/api/commands/SortedSetBaseCommands.java
Comment (Yury-Fridlyand) [reply java/client/src/main/java/glide/api/commands/SortedSetBaseCommands.java:910]: It is spotless trics, unfortunately

File: java/integTest/src/test/java/glide/SharedCommandTests.java
Comment (Yury-Fridlyand) [reply java/integTest/src/test/java/glide/SharedCommandTests.java:2406]: Two functions one command... oops

File: java/client/src/main/java/glide/api/BaseClient.java
Comment (acarbonetto) [reply java/client/src/main/java/glide/api/BaseClient.java:1081]: we can totally reorder them.

File: glide-core/src/client/value_conversion.rs
Comment (jonathanl-bq) [inline glide-core/src/client/value_conversion.rs:201]: If you're going to unwrap anyway, you can just use array indexing syntax.
```suggestion
                if array.len() == 2 && matches!(array[1], Value::Array(_)) =>
```

File: glide-core/src/client/value_conversion.rs
Comment (jonathanl-bq) [inline glide-core/src/client/value_conversion.rs:197]: nil or what? Should this just be "command returns nil"?

File: glide-core/src/client/value_conversion.rs
Comment (jonathanl-bq) [inline glide-core/src/client/value_conversion.rs:182]: nit: indentation seems a bit off here

File: java/client/src/main/java/glide/api/BaseClient.java
Comment (Yury-Fridlyand) [reply java/client/src/main/java/glide/api/BaseClient.java]: Done

File: glide-core/src/client/value_conversion.rs
Comment (Yury-Fridlyand) [reply glide-core/src/client/value_conversion.rs]: I trusted clippy/fmt to fix that...

Comment (jonathanl-bq) [reply glide-core/src/client/value_conversion.rs]: `clippy` doesn't care about formatting. It gives suggestions on how to write more idiomatic and performant code. `fmt` would be responsible for styling, but it's generally pretty lax compared to other formatters (like Spotless) about things like indentation because some people really hate that (https://news.ycombinator.com/item?id=10448667).

Comment (Yury-Fridlyand) [reply glide-core/src/client/value_conversion.rs]: TIL
Thanks for describing things

File: glide-core/src/client/value_conversion.rs
Comment (jonathanl-bq) [inline glide-core/src/client/value_conversion.rs:186]:
```suggestion
        // command returns nil or an array of 2 elements, where the second element is a map represented by a 2D array
```

**Review by jonathanl-bq — APPROVED** [review]

**Pull request merged**
